### PR TITLE
Introduction to what is backend and technologies

### PR DESCRIPTION
**Pull request by swighe**
# Pull Request Checklist

- [ ] Copy does not include typos or grammatical mistakes
- [ ] Confidential or company info has not been committed
- [ ] Semester-specific copy has been changed accordingly

### CODE REVIEW

File: notes/BACKEND.md [hunk 1]
```diff
@@ -1,3 +1,31 @@
 # Back-end Systems
 
-_Notes for this topic currently under development_
\ No newline at end of file
+Simply put, back-end systems cover aspects of development that don't involve any kind of user interface. This includes API's, scripts, libraries etc. that don't have a user interface, but run on the server side and provide a unified view to one or many consumers.
+
+## Difference between front-end and back-end systems
+Front-end systems interface with the user via a browser or an application. Some popular web technologies include HTML, CSS, and JavaScript, many frameworks like Bootstrap and Angular, as well as JavaScript libraries like jQuery, and CSS extensions like Sass and LESS.
+
+Typically, back-end systems run on the server side and serve front-end requests. This can be as simple as a static resource, or something like big data processes or restful API's that interface with data stores before returning the computed result. Some examples for back-end languages include Ruby, PHP, Java, C#, Python etc. with lots of supported frameworks like Django for Python, Sprint for Java, NodeJs, Symphony for Php etc.
```
Comment (jameslinjl) [inline notes/BACKEND.md]: Spring for Java

Comment (swighe) [reply notes/BACKEND.md]: Thanks fixed

File: notes/BACKEND.md [hunk 2]
```diff
@@ -1,3 +1,31 @@
 # Back-end Systems
 
-_Notes for this topic currently under development_
\ No newline at end of file
+Simply put, back-end systems cover aspects of development that don't involve any kind of user interface. This includes API's, scripts, libraries etc. that don't have a user interface, but run on the server side and provide a unified view to one or many consumers.
+
+## Difference between front-end and back-end systems
+Front-end systems interface with the user via a browser or an application. Some popular web technologies include HTML, CSS, and JavaScript, many frameworks like Bootstrap and Angular, as well as JavaScript libraries like jQuery, and CSS extensions like Sass and LESS.
+
+Typically, back-end systems run on the server side and serve front-end requests. This can be as simple as a static resource, or something like big data processes or restful API's that interface with data stores before returning the computed result. Some examples for back-end languages include Ruby, PHP, Java, C#, Python etc. with lots of supported frameworks like Django for Python, Spring for Java, NodeJs, Symphony for Php etc.
```
Comment (jameslinjl) [inline notes/BACKEND.md]: `run on the server side and serve front-end requests` the wording here might be a little confusing

Comment (jameslinjl) [inline notes/BACKEND.md]: a couple concrete examples would be helpful

File: notes/BACKEND.md [hunk 3]
```diff
@@ -1,3 +1,31 @@
...
+
+## API's and REST
```
Comment (jameslinjl) [inline notes/BACKEND.md]: no apostrophe

File: notes/BACKEND.md [hunk 4]
```diff
@@ -1,3 +1,31 @@
...
+### API
```
Comment (jameslinjl) [inline notes/BACKEND.md]: again, provide an example of an API as opposed to direct access to the data source, i think that will make the concept clearer

File: notes/BACKEND.md [hunk 5]
```diff
@@ -1,3 +1,31 @@
...
+An API is an application programming interface that abstracts the back-end details and executes business logic to represent information to the user in a consistent manner. API's help achieve the following:
+* Provide a defined interface for requests and responses to the caller
+* Abstract business logic and back-end layer (eg. databases, resources) from the caller. You can change anything without the caller knowing.
+* Security in the form of authentication, authorization, and preventing attacks like DOS via mechanisms like throttling.
+* Validation layer before requests hit the back-end
+* Abstract the front-end user facing interface/model from the data access layer model
+* Provide a single, uniform layer/experience to all users of the system with one-time coding in one place
+* Provide mechanisms for caching or speed up
+### RESTful API
+A RESTful API is an application program interface (API) that uses HTTP requests to GET, PUT, POST and DELETE data. REST stands for "representational state transfer". It is an architectural style and approach to communications often used in web services development. Components of a REST API are:
```
Comment (jameslinjl) [inline notes/BACKEND.md]: Why has REST been so popular?

File: notes/BACKEND.md [hunk 6]
```diff
@@ -1,3 +1,31 @@
...
+* URI: Uniform Resource Identifier
+* Protocol for communicaton: eg. HTTP
+* Data format for exchange: Eg. JSON, XML, protobuf etc.
+* Methods: Most common methods include GET, PUT, POST, DELETE with HTTP, other methods exist for CRUD operations.
+
+Other protocols exist like SOAP (Simple Object Access Protocol) which is more heavy weight but has added security, GraphQL which is a game changer that can solve problems that REST cannot.
+
+## Other examples of backend technologies
```
Comment (jameslinjl) [inline notes/BACKEND.md]: each of these could probably use a very very brief example

File: notes/BACKEND.md [hunk 7]
```diff
@@ -1,3 +1,31 @@
...
+* Big data processing in the background applying business logic, mapping and aggregating data into meaningful streams
+* Offline ETL's/scripts performing some business function
```
Comment (jameslinjl) [inline notes/BACKEND.md]: could you provide a `Useful Resources` section providing links for where students can get more information?

File: notes/BACKEND.md (hunk 2 above)
Comment (swighe) [reply notes/BACKEND.md]: Fixed.

File: notes/BACKEND.md (hunk 3 above)
Comment (swighe) [reply notes/BACKEND.md]: removed

File: notes/BACKEND.md (hunk 4 above)
Comment (swighe) [reply notes/BACKEND.md]: Done

File: notes/BACKEND.md [hunk 8]
```diff
@@ -1,3 +1,31 @@
 # Back-end Systems
 
-_Notes for this topic currently under development_
\ No newline at end of file
+Simply put, back-end systems cover aspects of development that don't involve any kind of user interface. This includes API's, scripts, libraries etc. that don't have a user interface, but run on the server side and provide a unified view to one or many consumers.
```
Comment (jmgreenberg) [inline notes/BACKEND.md]: Found this online. Found it to be very elegant way to explain int. 
>  back end refer to the separation of concerns between the presentation layer (front end), and the data access layer (back end) of a piece of software, or the physical infrastructure or hardware.

https://en.wikipedia.org/wiki/Front_and_back_ends

File: notes/BACKEND.md (hunk 5 above)
Comment (swighe) [reply notes/BACKEND.md]: Done

File: notes/BACKEND.md (hunk 6 above)
Comment (swighe) [reply notes/BACKEND.md]: Done.

File: notes/BACKEND.md (hunk 7 above)
Comment (swighe) [reply notes/BACKEND.md]: Done

File: notes/BACKEND.md (hunk 8 above)
Comment (swighe) [reply notes/BACKEND.md]: Addressed

File: notes/BACKEND.md [hunk 9]
```diff
@@ -1,3 +1,55 @@
 # Back-end Systems
 
-_Notes for this topic currently under development_
\ No newline at end of file
+This section adpated from Reference 1.
```
Comment (jameslinjl) [inline notes/BACKEND.md]: sp: adapted

applies to multiple places in doc

Comment (swighe) [reply notes/BACKEND.md]: fixed.

**Review by swighe — COMMENTED** [review]
typo

**Review by jameslinjl — APPROVED** [review]

**Pull request merged**